### PR TITLE
Build with go install

### PR DESCRIPTION
**Pull request by jawnsy**
Closes: #313

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -50,24 +42,34 @@ conmon:
...
+	$(GOPATH)/bin/ocid --config="" config --default > ocid.conf
 
 clean:
-	rm -f ocid.conf
-	rm -f ocic ocid
-	rm -f kpod
-	rm -f ${OCID_LINK}
+ifdef GOPATH
+	rm -rf $(GOPATH)/pkg
+	rm -f $(GOPATH)/bin/*
```
Comment (mrunalp) [inline Makefile]: We should only remove ocic, ocid and kpod instead of all binaries from GOPATH.

Comment (jawnsy) [inline Makefile]: Okay, will do. That's safer for sure :+1:

File: Makefile [hunk 2]
```diff
@@ -50,24 +42,34 @@ conmon:
...
-	./ocid --config="" config --default > ocid.conf
+	$(GOPATH)/bin/ocid --config="" config --default > ocid.conf
 
 clean:
-	rm -f ocid.conf
-	rm -f ocic ocid
-	rm -f kpod
-	rm -f ${OCID_LINK}
+ifdef GOPATH
+	rm -rf $(GOPATH)/pkg
```
Comment (mrunalp) [inline Makefile]: Same for the pkg. We will have to remove $GOPATH/pkg/"$OS_$ARCH"/github.com/kubernetes-incubator/cri-o/

**Comment by mrunalp** [comment]
Another issue I see while testing:
```
 make install.tools
go get github.com/vbatts/git-validation
go get github.com/alecthomas/gometalinter
gometalinter --install
Installing:
  aligncheck
  staticcheck
  varcheck
  lll
  gas
  gocyclo
  golint
  gosimple
  gotype
  ineffassign
  interfacer
  misspell
  deadcode
  goconst
  goimports
  unconvert
  dupl
  errcheck
  structcheck
  unused
can't load package: package github.com/HewlettPackard/gas: cannot find package "github.com/HewlettPackard/gas" in any of:
	/usr/lib/golang/src/github.com/HewlettPackard/gas (from $GOROOT)
	/home/mrunal/go/src/github.com/alecthomas/gometalinter/vendor/src/github.com/HewlettPackard/gas (from $GOPATH)
WARNING: failed to install one or more linters: exit status 1 (installing individually)
can't load package: package github.com/HewlettPackard/gas: cannot find package "github.com/HewlettPackard/gas" in any of:
	/usr/lib/golang/src/github.com/HewlettPackard/gas (from $GOROOT)
	/home/mrunal/go/src/github.com/alecthomas/gometalinter/vendor/src/github.com/HewlettPackard/gas (from $GOPATH)
WARNING: failed to install github.com/HewlettPackard/gas: exit status 1
gometalinter: error: failed to install the following linters: github.com/HewlettPackard/gas
make: *** [Makefile:167: .install.gometalinter] Error 1
```

**Comment by jawnsy** [comment]
Debugged @mrunalp's problem, turns out there was a `gometalinter` already installed in `/usr/bin`. We've decided to address the problem by installing tools to `GOPATH/bin` and searching in there instead, so that we always use a local copy of tools.

**Comment by mheon** [comment]
Builds and tests fine on my system

**Comment by mrunalp** [comment]
Tests work here, too (except #322).

**Comment by mrunalp** [comment]
LGTM

**Comment by mrunalp** [comment]
@runcom @rhatdan @sameo @cyphar PTAL

**Comment by mrunalp** [comment]
This should fix #311 as well.

**Comment by runcom** [comment]
Awesome, great job @jawnsy

**Comment by runcom** [comment]
I'll run tests locally (given the known ping host failure) and merge in a while.

**Comment by runcom** [comment]
everything works fine on my system :) (except that now I have to run `sudo GOPATH=$GOPATH make localintegration` to run tests, but that's fine)

**Comment by runcom** [comment]
Thanks @jawnsy

**Comment by jawnsy** [comment]
Thanks for the quick review & merge, @runcom!

**Comment by jawnsy** [comment]
@runcom About the `GOPATH`, you can use `sudo -E` to preserve your environment variables (including `GOPATH`). In a follow-up, I'll look at building the test binary as non-root (looking at the docs for `go test` it looks possible) and sudo running the test program.

**Comment by cyphar** [comment]
*sigh* Now everything breaks on my system and I can no longer build `cri-o`. Was the old system really so broken that it needed to be replaced with another system that doesn't work as well as the old one? Not to mention the fact that it was broken after two or three days of review -- given that it's holiday season maybe it's a bit odd to expect that nobody was too busy to read their backlog of ~700 emails?

**Comment by cyphar** [comment]
In particular, this thing that I don't understand is the logic behind this change -- `go install` combines the compilation and installation of binaries. I don't know of a single distribution where this is how the build process works -- `deb`, `rpm` and `PKGBUILD` all have separate "build" and "install" sections.

So the real question is, who is the `Makefile` targeted at?

**Comment by cyphar** [comment]
Not to mention that as far as I can tell, `go install` will install the static libraries and I cannot find where it's shoved the compiled binaries. Remind me how this is better again?

**Comment by cyphar** [comment]
Oh, and I currently cannot find where the binaries are installed (it's not in `$GOPATH/pkg`)... **THIS** is why distributions don't use the `Makefile`s of upstream Go projects -- because Go makes it ridiculously easy to break the build system.

**Comment by cyphar** [comment]
After reading the discussion in #313 I can see why you wanted this @jawnsy. However here are a few points:

1. The `vendor/src` -> `vendor` change means you cannot compile with older versions of Go. And also now certain `GOPATH` tricks no longer work (and no, you can't fix them with symlinks). Okay, maybe we don't care about this.

2. `go install` is a touch too magical for me. In particular, it doesn't produce single binaries -- it attempts to install things as well (in essentially random paths). And that then doesn't work for me _at all_ (it just generates `.a` files), not to mention that it breaks the package building process of all distributions that I'm aware of.

So we really need to consider what purpose our `Makefile` serves. If you want to use `go install`, that's fine -- but lets not break the `Makefile` in order to make it so that distributions can no longer use our `Makefile`...

**Pull request merged**